CTS: RESTful WS TCK failing when MBW attempts to write a JsonString or JsonNumber

From conversation in slack with @brideck :

> For jsonp, I see:
```
Caused by: jakarta.ws.rs.ProcessingException: RESTEASY003215: could not find writer for content-type application/json type: org.glassfish.json.JsonStringImpl
	at org.jboss.resteasy.core.interception.jaxrs.ClientWriterInterceptorContext.throwWriterNotFoundException(ClientWriterInterceptorContext.java:50)
	at org.jboss.resteasy.core.interception.jaxrs.AbstractWriterInterceptorContext.getWriter(AbstractWriterInterceptorContext.java:302)
	at org.jboss.resteasy.core.interception.jaxrs.AbstractWriterInterceptorContext.syncProceed(AbstractWriterInterceptorContext.java:240)
	at org.jboss.resteasy.core.interception.jaxrs.AbstractWriterInterceptorContext.proceed(AbstractWriterInterceptorContext.java:224)
	at org.jboss.resteasy.client.jaxrs.internal.ClientInvocation.writeRequestBody(ClientInvocation.java:440)
	...
```

> Note that it also fails for `org.glassfish.json.JsonNumberImpl$JsonLongNumber`

I believe this is because our JSON-P Provider only checks whether the type is a JsonObject or JsonArray, when it should check whether the type is a JsonValue (which is the super-interface for JsonObject, JsonArray, JsonString, JsonNumber, etc.).

--- a/dev/io.openliberty.org.jboss.resteasy.common/src/io/openliberty/restfulWS/providers/JsonPProvider.java
+++ b/dev/io.openliberty.org.jboss.resteasy.common/src/io/openliberty/restfulWS/providers/JsonPProvider.java
@@ -17,14 +17,11 @@ import java.lang.annotation.Annotation;
 import java.lang.reflect.Type;
 import java.security.AccessController;
 import java.security.PrivilegedAction;
-import java.util.Iterator;
-import java.util.ServiceLoader;
 
-import javax.json.JsonArray;
+
 import javax.json.JsonException;
-import javax.json.JsonObject;
 import javax.json.JsonReader;
-import javax.json.JsonStructure;
+import javax.json.JsonValue;
 import javax.json.JsonWriter;
 import javax.json.spi.JsonProvider;
 import javax.ws.rs.BadRequestException;
@@ -74,9 +71,7 @@ public class JsonPProvider implements MessageBodyReader<Object>, MessageBodyWrit
 
     @Override
     public boolean isWriteable(Class<?> type, Type genericType, Annotation[] annotations, MediaType mediaType) {
-        return JsonStructure.class.isAssignableFrom(type) ||
-               JsonArray.class.isAssignableFrom(type) ||
-               JsonObject.class.isAssignableFrom(type);
+        return JsonValue.class.isAssignableFrom(type);
     }
 
     @Override
@@ -90,7 +85,7 @@ public class JsonPProvider implements MessageBodyReader<Object>, MessageBodyWrit
         try {
             writer = this.jsonProvider.createWriter(entityStream);
             if (writer != null) {
-                writer.write((JsonStructure) obj);
+                writer.write((JsonValue) obj);
             }
         } catch (Throwable e) {
             //ignore
@@ -107,9 +102,7 @@ public class JsonPProvider implements MessageBodyReader<Object>, MessageBodyWrit
 
     @Override
     public boolean isReadable(Class<?> type, Type genericType, Annotation[] annotations, MediaType mediaType) {
-        return JsonStructure.class.isAssignableFrom(type) ||
-               JsonArray.class.isAssignableFrom(type) ||
-               JsonObject.class.isAssignableFrom(type);
+        return JsonValue.class.isAssignableFrom(type);
     }
 
     @FFDCIgnore(value = { Throwable.class })

--- a/dev/io.openliberty.org.jboss.resteasy.common/test/io/openliberty/restfulWS/providers/JsonPProviderTest.java
+++ b/dev/io.openliberty.org.jboss.resteasy.common/test/io/openliberty/restfulWS/providers/JsonPProviderTest.java
@@ -1,0 +1,101 @@
+/*******************************************************************************
+ * Copyright (c) 2020 IBM Corporation and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *     IBM Corporation - initial API and implementation
+ *******************************************************************************/
+package io.openliberty.restfulWS.providers;
+
+import static javax.ws.rs.core.MediaType.APPLICATION_JSON_TYPE;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertTrue;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.InputStream;
+import java.lang.annotation.Annotation;
+
+import javax.json.Json;
+import javax.json.JsonArray;
+import javax.json.JsonObject;
+import javax.json.JsonString;
+import javax.ws.rs.core.MultivaluedHashMap;
+
+import org.junit.Test;
+
+public class JsonPProviderTest {
+
+    JsonPProvider provider = new JsonPProvider();
+
+    @Test
+    public void testMBRforJsonObject() throws Exception {
+        InputStream stream = new ByteArrayInputStream("{\"foo\": \"bar\"}".getBytes());
+        assertTrue(provider.isReadable(JsonObject.class, JsonObject.class, new Annotation[] {}, APPLICATION_JSON_TYPE));
+        JsonObject obj = (JsonObject) provider.readFrom(Object.class, JsonObject.class, new Annotation[] {}, 
+                                                        APPLICATION_JSON_TYPE, new MultivaluedHashMap<>(), stream);
+        assertTrue(obj.containsKey("foo"));
+        assertEquals("bar", obj.getString("foo"));
+    }
+
+    @Test
+    public void testMBWforJsonObject() throws Exception {
+        ByteArrayOutputStream stream = new ByteArrayOutputStream();
+        JsonObject obj = Json.createObjectBuilder().add("foo", "bar").build();
+        assertTrue(provider.isWriteable(obj.getClass(), obj.getClass(), new Annotation[] {}, APPLICATION_JSON_TYPE));
+        provider.writeTo(obj, obj.getClass(), obj.getClass(), new Annotation[] {}, APPLICATION_JSON_TYPE, new MultivaluedHashMap<>(), stream);
+        assertEquals("{\"foo\":\"bar\"}", new String(stream.toByteArray()));
+    }
+
+    @Test
+    public void testMBRforJsonArray() throws Exception {
+        InputStream stream = new ByteArrayInputStream("[\"foo\", \"bar\", \"baz\"]".getBytes());
+        assertTrue(provider.isReadable(JsonArray.class, JsonArray.class, new Annotation[] {}, APPLICATION_JSON_TYPE));
+        JsonArray arr = (JsonArray) provider.readFrom(Object.class, JsonArray.class, new Annotation[] {}, 
+                                                      APPLICATION_JSON_TYPE, new MultivaluedHashMap<>(), stream);
+        assertNotNull(arr);
+        assertEquals(3, arr.size());
+        assertEquals("foo", ((JsonString)arr.get(0)).getString());
+        assertEquals("bar", ((JsonString)arr.get(1)).getString());
+        assertEquals("baz", ((JsonString)arr.get(2)).getString());
+    }
+
+    @Test
+    public void testMBWforJsonArray() throws Exception {
+        ByteArrayOutputStream stream = new ByteArrayOutputStream();
+        JsonArray arr = Json.createArrayBuilder().add("foo").add("bar").add("baz").build();
+        assertTrue(provider.isWriteable(arr.getClass(), arr.getClass(), new Annotation[] {}, APPLICATION_JSON_TYPE));
+        provider.writeTo(arr, arr.getClass(), arr.getClass(), new Annotation[] {}, APPLICATION_JSON_TYPE, new MultivaluedHashMap<>(), stream);
+        assertEquals("[\"foo\",\"bar\",\"baz\"]", new String(stream.toByteArray()));
+    }
+
+    @Test
+    public void testMBWforJsonString() throws Exception {
+        ByteArrayOutputStream stream = new ByteArrayOutputStream();
+        JsonString str = new JsonString() {
+
+            @Override
+            public ValueType getValueType() {
+                return ValueType.STRING;
+            }
+
+            @Override
+            public CharSequence getChars() {
+                return "Hello World!";
+            }
+
+            @Override
+            public String getString() {
+                return "Hello World!";
+            }
+            
+        };
+        assertTrue(provider.isWriteable(str.getClass(), str.getClass(), new Annotation[] {}, APPLICATION_JSON_TYPE));
+        provider.writeTo(str, str.getClass(), str.getClass(), new Annotation[] {}, APPLICATION_JSON_TYPE, new MultivaluedHashMap<>(), stream);
+        assertEquals("\"Hello World!\"", new String(stream.toByteArray()));
+    }
+}

